data_structures/raster_source.cpp

```diff
@@ -0,0 +1,183 @@
+ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+*/
+
+#include "raster_source.hpp"
+
+#include "../util/simple_logger.hpp"
+#include "../util/timing_util.hpp"
+
+#include <osrm/coordinate.hpp>
+
+#include <sstream>
+#include <cmath>
+#include <iostream>
+
+RasterSource::RasterSource(RasterGrid _raster_data,
+                           std::size_t _width,
+                           std::size_t _height,
+                           int _xmin,
+                           int _xmax,
+                           int _ymin,
+                           int _ymax)
+    : xstep(calcSize(_xmin, _xmax, _width)), ystep(calcSize(_ymin, _ymax, _height)),
+      raster_data(_raster_data), width(_width), height(_height), xmin(_xmin), xmax(_xmax),
+      ymin(_ymin), ymax(_ymax)
+{
+    BOOST_ASSERT(xstep != 0);
+    BOOST_ASSERT(ystep != 0);
+}
+
+float RasterSource::calcSize(double min, double max, unsigned count) const
+{
+    BOOST_ASSERT(count > 0);
+    return (max - min) / count;
+}
+
+// Query raster source for nearest data point
+RasterDatum RasterSource::getRasterData(const float lon, const float lat) const
+{
+    int _lon = static_cast<int>(lon * COORDINATE_PRECISION);
+    int _lat = static_cast<int>(lat * COORDINATE_PRECISION);
+
+    if (_lon < xmin || _lon > xmax || _lat < ymin || _lat > ymax)
+    {
+        return {};
+    }
+
+    float xthP = (_lon - xmin) / xstep;
+    int xth = ((xthP - floor(xthP)) > (xstep / 2.) ? floor(xthP) : ceil(xthP));
+
+    float ythP = (ymax - _lat) / ystep;
+    int yth = ((ythP - floor(ythP)) > (ystep / 2.) ? floor(ythP) : ceil(ythP));
+
+    return {raster_data(xth, yth)};
+}
+
+// Query raster source using bilinear interpolation
+RasterDatum RasterSource::getRasterInterpolate(const float lon, const float lat) const
+{
+    int _lon = static_cast<int>(lon * COORDINATE_PRECISION);
+    int _lat = static_cast<int>(lat * COORDINATE_PRECISION);
+
+    if (_lon < xmin || _lon > xmax || _lat < ymin || _lat > ymax)
+    {
+        return {};
+    }
+
+    unsigned xthP = (_lon - xmin) / xstep;
+    unsigned ythP = (ymax - _lat) / ystep;
+    int top = floor(ythP);
+    int bottom = ceil(ythP);
+    int left = floor(xthP);
+    int right = ceil(xthP);
+
+    float x = (_lon - left * xstep + xmin) / xstep;
+    float y = (ymax - top * ystep - _lat) / ystep;
+    float x1 = 1.0 - x;
+    float y1 = 1.0 - y;
+
+    return {static_cast<std::int32_t>(
+        raster_data(left, top) * (x1 * y1) + raster_data(right, top) * (x * y1) +
+        raster_data(left, bottom) * (x1 * y) + raster_data(right, bottom) * (x * y))};
```

Comment:
This looks like a bilinear interpolation and probably is. Would be great to derive it from the code at first sight.
Maybe introduce some intermediary variables and give them appropriate names.